Multiplexing Support

averycblack opened the pull request:
Related Issue: https://github.com/acidanthera/bugtracker/issues/1584

Some HP devices with a trackpoint use a mode on the i8042 controller which allows 4 multiplexed devices to connect where usually only one Auxiliary device would connect. Instead of using PS2 passthrough, HP will connect the trackpoint to one port on the controller, and the trackpad to another. With VoodooRMI, the trackpad needed to be disabled while still allowing the trackpoint to work over PS2. To help with this, this PR puts the controller into the multiplexed mode when it is detected.

Some other things done in this PR to help with the above:
* Create either 1 or 4 mouse nubs depending on if the controller is multiplexed or not.
* Removed the mouse specific commands, and instead have the controller decide when to transmit to the mouse based off of the port.
* Removed the "WriteCommandPort" command. This was only used to tell the controller to transmit to the mouse.
* Removes a workaround which allowed VoodooPS2Synaptic to attach to type 0x46 trackpads.

I believe that the 0x46 type workaround was originally put in for an old HP device with a Synaptic trackpoint in a similar multiplexed device. I have no way to test, but my guess is that the type is read from the trackpoint instead of the trackpad. This workaround broke the device I was having tested, as the trackpad reported 0x47 while the trackpoint reported 0x46. This made VoodooPS2Synaptics attach to the trackpoint, which it isn't compatible with.

When multiplexed, the device looks like this in ioreg:
![image](https://user-images.githubusercontent.com/13190749/129487265-7ec173ab-9397-453d-a1d1-6965fccd4670.png)

Todo:
- [x] Add multiplexing support to Debugger mode
- [x] Add multiplexing support when using HANDLE_INTERRUPT_DATA_LATER
- [x] Re-add watchdog mode (accidentally got rid of it). Was this ever used?

averycblack committed:
Muxed first draft, works in non-muxed mode

averycblack committed:
Make analyzer happy

averycblack committed:
Remove check for device type 0x46
This prevented the trackstick from working as it was seen as a trackpad

averycblack commented:
Workaround is first mentioned here (search for 0x46): https://github.com/RehabMan/OS-X-Voodoo-PS2-Controller

This project also is a giant mess of 2 space, 4 space, and tab indentation all in the same file. I'll probably try to fix that in a different PR.

averycblack committed:
interrupt later and debugger support

averycblack committed:
Add back in watchdog mode, check null in interrupt in ApplePS2Device
_interruptSource can't be null. packetAction can be called directly when using HANDLE_INTERRUPT_DATA_LATER

averycblack committed:
Listen on muxed port on reset

vit9696 commented on VoodooPS2Controller/ApplePS2Device.h:
```diff
@@ -27,11 +27,12 @@
 #ifdef DEBUG_MSG
 #define DEBUG_LOG(args...)  do { IOLog(args); } while (0)
-#define INFO_LOG(args...)  do { IOLog(args); IOSleep(1000); } while (0)
+#define INFO_LOG(args...)  do { IOLog(args); } while (0)
```
This delay was done to avoid missed messages during debugging from what I remember. Did you get a kernel panic with it? If so, maybe we should separate logs where we can delay and where we cannot (i.e. interrupt delays).

cc @usr-sse2, remind me whether I am right and please express your opinion here.

averycblack replied:
I'll revert it. I got annoyed with the delay while loading/unloading it in Catalina, though I should've remembered to add it back. 👍

vit9696 commented on VoodooPS2Controller/ApplePS2Device.h:
```diff
@@ -617,6 +611,15 @@ class EXPORT ApplePS2Device : public IOService
+    IOWorkLoop * _workloop {nullptr};
+    IOInterruptEventSource * _interruptSource {nullptr};
+    
+    OSObject* _client;
```
Please nullptr this.

vit9696 commented on VoodooPS2Controller/VoodooPS2Controller.cpp:
```diff
@@ -158,20 +158,19 @@ void ApplePS2Controller::onWatchdogTimer()
-    
-    bool wakeMouse = false;
-    bool wakeKeyboard = false;
+    bool wakePort[kPS2MaxIdx] = {false, false, false, false, false, false};
```
bool wakePort[kPS2MaxIdx] {};

Otherwise it depends on kPS2MaxIdx not reducing.

averycblack replied:
Do you mind explaining this? Not quite sure what you mean by reducing.

vit9696 replied:
Reducing the constant value. If you make it less, the code will not compile. If you increase it, it will also give a false impression of something not being initialised, which may confuse somebody. Writing the code like this gives you a named constant but forces you to remember its value implicitly, which is not right.

vit9696 commented on VoodooPS2Controller/VoodooPS2Controller.cpp:
```diff
@@ -445,6 +456,34 @@ void ApplePS2Controller::resetController(void)
+        return false;
+    writeCommandPort(kCP_WriteMouseOutputBuffer);
+    writeDataPort(0x56);
+    if (readDataPort(kPS2AuxIdx) != 0x56)
```
Can we at least try to name these magic constants?

averycblack replied:
The constants aren't labeled in the Linux source, and I haven't found any other documentation. I atleast put them in one place though.

vit9696 commented on VoodooPS2Controller/VoodooPS2Controller.cpp:
```diff
@@ -575,30 +608,48 @@ bool ApplePS2Controller::start(IOService * provider)
+             !_devices[i]->init(i)            ||
+             !_devices[i]->attach(this) )
+        {
+          OSSafeReleaseNULL(_devices[i]);
```
Should not it abort on failure?

vit9696 commented on VoodooPS2Controller/VoodooPS2Controller.cpp:
```diff
@@ -575,30 +608,48 @@ bool ApplePS2Controller::start(IOService * provider)
   {
-	  OSSafeReleaseNULL(_mouseDevice);
-	  OSSafeReleaseNULL(_interruptSourceMouse);
+    for (int i = kPS2MuxIdx; i < kPS2MaxIdx; i++)
```
Also, why not size_t here and in many other places.

vit9696 commented on VoodooPS2Controller/VoodooPS2Controller.cpp:
```diff
@@ -798,40 +844,25 @@ void ApplePS2Controller::uninstallInterruptAction(PS2DeviceType deviceType)
-    _packetActionMouse = NULL;
-    _interruptTargetMouse->release();
-    _interruptTargetMouse = 0;
+    _interruptInstalledMouse--;
```
Put an assert here that it is not already 0. The previous code does check this, if you believe it is impossible, please at least add a debug check.

vit9696 commented on VoodooPS2Controller/VoodooPS2Controller.cpp:
```diff
@@ -1088,74 +1116,53 @@ void ApplePS2Controller::processRequest(PS2Request * request)
-        request->commands[index].inOrOut = readDataPort(deviceMode);
-        break;
-            
-      case kPS2C_ReadMouseDataPortAndCompare:
```
Why do we no longer need these commands even for setups without mux?

averycblack replied:
The only difference between talking to a mouse compared to a keyboard is that an additional command (kCP_TransmitToMouse) is written to the command port. Talking to a muxed device also requires an additional command (kCP_TransmitToMuxedMouse). Otherwise the logic is the same between the keyboard and mouse.

It made more sense to me to write to the command port based off of the port number of the request. This made it so that the trackpad drivers didn't care about the difference between a mouse and keyboard command. There shouldn't be any reason that the trackpad should be talking to the keyboard port anyways.

vit9696 commented on VoodooPS2Trackpad/VoodooPS2SynapticsTouchPad.cpp:
```diff
@@ -190,7 +190,7 @@ ApplePS2SynapticsTouchPad* ApplePS2SynapticsTouchPad::probe(IOService * provider
     else
     {
         INFO_LOG("VoodooPS2Trackpad: Identify bytes = { 0x%x, 0x%x, 0x%x }\n", buf3[0], buf3[1], buf3[2]);
-        if (0x46 != buf3[1] && 0x47 != buf3[1])
```
Is this the hack you mentioned you removed? CC @usr-sse2 what do you think about it?

vit9696 reviewed (commented):
Thank you for this. I commented briefly, but it is hard to follow all the changes and the only real state can be understood with testing. I guess we can merge this and see what reports will we get, as long as you keep an eye on the comings.

averycblack committed:
Apply code suggestions
Use size_t for port
Assert if _interruptInstalledMouse becomes less than zero
nullptr initializer for _client
Add delay back in for INFO_LOG

averycblack commented:
> 
> 
> Thank you for this. I commented briefly, but it is hard to follow all the changes and the only real state can be understood with testing. I guess we can merge this and see what reports will we get, as long as you keep an eye on the comings.

I'm not expecting any regresions, I've been testing on my X1 Extreme which isn't muxed. I'll keep an eye out though for issues on the bugtracker if this gets merged.

vit9696 commented:
Merging this early to force testing. @usr-sse2 it will help if you can still comment post-mortem.

vit9696 commented:
```
Hi,

Please find the latest report on new defect(s) introduced to acidanthera/VoodooPS2 found with Coverity Scan.

2 new defect(s) introduced to acidanthera/VoodooPS2 found with Coverity Scan.


New defect(s) Reported-by: Coverity Scan
Showing 2 of 2 defect(s)


** CID 338519: Control flow issues (NO_EFFECT)
/VoodooPS2Controller/VoodooPS2Controller.cpp: 976 in ApplePS2Controller::submitRequestAndBlock(PS2Request *)()


________________________________________________________________________________________________________
*** CID 338519: Control flow issues (NO_EFFECT)
/VoodooPS2Controller/VoodooPS2Controller.cpp: 976 in ApplePS2Controller::submitRequestAndBlock(PS2Request *)()
970
971 // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
972
973 void ApplePS2Controller::submitRequestAndBlock(PS2Request * request)
974 {
975 // Make sure that the request is in bounds
>>> CID 338519: Control flow issues (NO_EFFECT)
>>> This less-than-zero comparison of an unsigned value is never true. "request->port < kPS2KbdIdx".
976 if (request->port >= kPS2MaxIdx || request->port < kPS2KbdIdx)
977 return;
978
979 // Check that we only read from ports which are active in mux mode
980 if ((_mux_present && request->port == kPS2AuxIdx) ||
981 (!_mux_present && request->port > kPS2AuxIdx))

** CID 338518: Control flow issues (NO_EFFECT)
/VoodooPS2Controller/VoodooPS2Controller.cpp: 951 in ApplePS2Controller::submitRequest(PS2Request *)()


________________________________________________________________________________________________________
*** CID 338518: Control flow issues (NO_EFFECT)
/VoodooPS2Controller/VoodooPS2Controller.cpp: 951 in ApplePS2Controller::submitRequest(PS2Request *)()
945
946 // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
947
948 bool ApplePS2Controller::submitRequest(PS2Request * request)
949 {
950 // Make sure that the request is in bounds
>>> CID 338518: Control flow issues (NO_EFFECT)
>>> This less-than-zero comparison of an unsigned value is never true. "request->port < kPS2KbdIdx".
951 if (request->port >= kPS2MaxIdx || request->port < kPS2KbdIdx)
952 return false;
953
954 // Check that we only read from ports which are active in mux mode
955 if ((_mux_present && request->port == kPS2AuxIdx) ||
956 (!_mux_present && request->port > kPS2AuxIdx))

averycblack commented:
Do those checks need to exist? Now that I think about it, the only way those errors would occur is if the MouseDevice/KeyboardDevice nub was initialized with invalid ports. It should be basically impossible.

I fixed the defects here: https://github.com/acidanthera/VoodooPS2/pull/39
